Game/Source/GDKShooter/Private/Game/OnlineServicesLibrary.cpp

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "OnlineServicesLibrary.h"
+
+#include "PlayFab.h"
+#include "Core/PlayFabAdminAPI.h"
+#include "Runtime/Online/HTTP/Public/Http.h"
+#include "JsonUtilities.h"
+
+void UOnlineServicesLibrary::SetupPlayFabSettings()
+{
+	// Get the CDO so we can read the config values.
+	UOnlineServicesLibrary* DefaultObject = Cast<UOnlineServicesLibrary>(UOnlineServicesLibrary::StaticClass()->GetDefaultObject());
+
+	PlayFabAdminPtr AdminApi = IPlayFabModuleInterface::Get().GetAdminAPI();
+	AdminApi->SetTitleId(DefaultObject->PlayFabTitleID);
+	AdminApi->SetDevSecretKey(DefaultObject->PlayFabSecretKey);
+}
+
+void UOnlineServicesLibrary::SendPOSTRequest(FString Endpoint, FString Path, TSharedPtr<FJsonObject> Content, TFunction<void(const bool, TSharedPtr<FJsonObject>)> Callback)
+{
+	// Get the CDO so we can read the config values.
+	UOnlineServicesLibrary* DefaultObject = Cast<UOnlineServicesLibrary>(UOnlineServicesLibrary::StaticClass()->GetDefaultObject());
+
+	FString OutputString;
+	TSharedRef<TJsonWriter<TCHAR>> JsonWriter = TJsonWriterFactory<>::Create(&OutputString);
+	FJsonSerializer::Serialize(Content.ToSharedRef(), JsonWriter);
+
+	TSharedRef<IHttpRequest> HttpRequest = FHttpModule::Get().CreateRequest();
+
+	HttpRequest->SetVerb("POST");
+	HttpRequest->SetHeader("Content-Type", "application/json");
+	HttpRequest->SetURL(*FString::Printf(TEXT("http://%s.endpoints.%s.cloud.goog/v1/%s"), *Endpoint, *DefaultObject->GoogleCloudProjectID, *Path));
+
+	UE_LOG(LogTemp, Log, TEXT("Sent POST request to %s: %s"), *HttpRequest->GetURL(), *OutputString);
+
+	HttpRequest->SetContentAsString(OutputString);
+
+	HttpRequest->OnProcessRequestComplete().BindLambda([Callback](FHttpRequestPtr Request, FHttpResponsePtr Response, bool bWasSuccessful)
```

Comment:
There are a few uses of `BindLambda` (In the OnlineServicesLibrary and the PartySystem), which could potentially be a crash risk? We had a difficult to track bug earlier due to `this` being null in the callback of a lambda (due to some unreal garbage collection). Possibly we should be using `BindWeakLambda` instead? (@Vatyx)

Reply:
So the requests in the OnlineServicesLibrary capture `Callback` which is a TFunction that often captures `this`. I believe this still has the risk of crashing if `this` gets cleaned up. Seems fine without it, but because there is the risk, I will probably look into a safer alternative later today.